pkg/controller/resourceclaim/controller.go

```diff
@@ -0,0 +1,472 @@
+
+	// The deletedObjects cache keeps track of Pods for which we know that
+	// they have existed and have been removed. For those we can be sure
+	// that a ReservedFor entry needs to be removed.
+	deletedObjects *uidCache
+}
+
+const (
+	claimKeyPrefix = "claim:"
+	podKeyPrefix   = "pod:"
+)
+
+// NewController creates a ResourceClaim controller.
+func NewController(
+	kubeClient clientset.Interface,
+	podInformer v1informers.PodInformer,
+	claimInformer resourcev1alpha1informers.ResourceClaimInformer,
+	templateInformer resourcev1alpha1informers.ResourceClaimTemplateInformer) (*Controller, error) {
+
+	ec := &Controller{
+		kubeClient:      kubeClient,
+		podLister:       podInformer.Lister(),
+		podIndexer:      podInformer.Informer().GetIndexer(),
+		podSynced:       podInformer.Informer().HasSynced,
+		claimLister:     claimInformer.Lister(),
+		claimsSynced:    claimInformer.Informer().HasSynced,
+		templateLister:  templateInformer.Lister(),
+		templatesSynced: templateInformer.Informer().HasSynced,
+		queue:           workqueue.NewNamedRateLimitingQueue(workqueue.DefaultControllerRateLimiter(), "resource_claim"),
+		deletedObjects:  newUIDCache(maxUIDCacheEntries),
+	}
+
+	metrics.RegisterMetrics()
+
+	eventBroadcaster := record.NewBroadcaster()
+	eventBroadcaster.StartLogging(klog.Infof)
+	eventBroadcaster.StartRecordingToSink(&v1core.EventSinkImpl{Interface: kubeClient.CoreV1().Events("")})
+	ec.recorder = eventBroadcaster.NewRecorder(scheme.Scheme, v1.EventSource{Component: "resource_claim"})
+
+	if _, err := podInformer.Informer().AddEventHandler(cache.ResourceEventHandlerFuncs{
+		AddFunc: func(obj interface{}) {
+			ec.enqueuePod(obj, false)
+		},
+		UpdateFunc: func(old, updated interface{}) {
+			ec.enqueuePod(updated, false)
+		},
+		DeleteFunc: func(obj interface{}) {
+			ec.enqueuePod(obj, true)
+		},
+	}); err != nil {
+		return nil, err
+	}
+	if _, err := claimInformer.Informer().AddEventHandler(cache.ResourceEventHandlerFuncs{
+		AddFunc: ec.onResourceClaimAddOrUpdate,
+		UpdateFunc: func(old, updated interface{}) {
+			ec.onResourceClaimAddOrUpdate(updated)
+		},
+		DeleteFunc: ec.onResourceClaimDelete,
+	}); err != nil {
+		return nil, err
+	}
+	if err := ec.podIndexer.AddIndexers(cache.Indexers{podResourceClaimIndex: podResourceClaimIndexFunc}); err != nil {
+		return nil, fmt.Errorf("could not initialize ResourceClaim controller: %w", err)
+	}
+
+	return ec, nil
+}
+
+func (ec *Controller) enqueuePod(obj interface{}, deleted bool) {
+	if d, ok := obj.(cache.DeletedFinalStateUnknown); ok {
+		obj = d.Obj
+	}
+	pod, ok := obj.(*v1.Pod)
+	if !ok {
+		// Not a pod?!
+		return
+	}
+
+	if deleted {
+		ec.deletedObjects.Add(pod.UID)
+	}
+
+	if len(pod.Spec.ResourceClaims) == 0 {
+		// Nothing to do for it at all.
+		return
+	}
+
+	// Release reservations of a deleted or completed pod?
+	if deleted ||
+		pod.DeletionTimestamp != nil && pod.DeletionGracePeriodSeconds != nil && *pod.DeletionGracePeriodSeconds == 0 ||
```

Comment:
Adding `pod.DeletionTimestamp != nil && pod.spec.NodeName == ""` - makes sense: that covers pods that got deleted before scheduling, but aren't going away (for example, because of a finalizer).

But we still need something for "pod got scheduled to a node and then deleted, with finalizer preventing removal". Can we rely on kubelet setting the phase in this case, i.e. is that covered by `pod.Status.Phase == v1.PodFailed || pod.Status.Phase == v1.PodSucceeded`?

It seems so. Okay, let's use that expression.

Reply:
I don't know for sure if kubelet would mark it as failed. Maybe you can check if it's Pending?